criu/sk-tcp.c

```diff
@@ -428,7 +430,7 @@ int restore_one_tcp(int fd, struct inet_sk_info *ii)
 
 	pr_info("Restoring TCP connection\n");
 
-	if (opts.tcp_close && ii->ie->state != TCP_LISTEN && ii->ie->state != TCP_CLOSE) {
+	if (dump_with_tcp_close | (opts.tcp_close && ii->ie->state != TCP_LISTEN && ii->ie->state != TCP_CLOSE)) {
```

Comment:
This looks wrong. It looks like it should be:

if ((dump_with_tcp_close || opts.tcp_close) && ii->ie->state != TCP_LISTEN && ii->ie->state != TCP_CLOSE) {

but restore_one_tcp is never called for listening sockets and I am not sure about TCP_CLOSE too...

Reply:
Yes, I see that TCP_LISTEN socket cannot reach here too. However, TCP_CLOSE sockets can. Without applying the patch, my added zdtm test will fail with the same error as the referenced issue `No tcp-stream-XXXX.img image`. This is because when `--tcp-close` is used when checkpointing, we skip dumping TCP for all TCP states. However, when restoring we still try to restore TCP_CLOSE state. So in my patch, I will skip restore TCP_CLOSE if the checkpoint happens with `--tcp-close`, the idea from @rst0git is great too as his idea is keep dumping TCP state in TCP_CLOSE case.

Reply:
@avagin Looks like `(dump_with_tcp_close || opts.tcp_close)` should be equal to `opts.tcp_close` in restore_one_tcp due to check in check_img_inventory.